rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+
+After the build, the runtime dependencies file is generated in-memory (if there are dependencies), the `$cas` is calculated including the runtime dependencies buffer, any occurences of the build path are replaced with `/var/lib/nix/$cas`, and the `$cas.m` file is written (if there are dependencies).
+
+If there undetected build path references, they might cause the finished entry to work incorrectly, and they will cause `$cas` to differ on every build of `$out`. This must be handled on a case-by-case basis. Perhaps we'll need pluggable hash rewriters.
+
+The build can happen by a sandboxing build daemon like `nix-build`, but that is not a requirement.
+
+After preparing, the metadata for the build is added to the user's Trust DB.
+
+### A note on `$cas` calculation
+
+The content hash has to be built up recursively, as described elsewhere. There are some caveats:
+
+- When the build path `$build/$out` is encountered, it should be replaced with `/var/lib/nix/$placeholder`. `$placeholder` is a fixed string used in all `$cas` calculations.
+- When `$out` is encountered but not `$build/$out`, it should be replaced with `$placeholder`.
+- When `$build` is encountered but not `$out`, it is a possible problem and should be noted.
+- The paths of files and directories should be included relative to `$build` (so `$build/$out` is `/$out`), but not in the file contents or symlink targets.
+- If there is metadata, the file should be included in the calculation as if it was named `_meta.m`, stored next to `$out`.
+
+Changing the entry to its `$cas` name means:
+
+- Rename `$out` to `$cas`.
+- Replace the string `$build/$out` with `/var/lib/nix/$cas`.
+- Replace the string `$out` with `$cas`.
+- If there is metadata, write the contents to `$cas.m`.
+
+When validating:
+
+- The paths of files and directories should be included relative to `/var/lib/nix`.
+- If `$cas.m` exists, it should be included in the calculation as if it was named `_meta.m`.
+- When the string `$cas` is encountered, it should be replaced with `$placeholder`.
+
+### Installation
+
+We use rename semantics to provide atomic installations. Prepared `$cas` entries are moved to their final location with a `rename` call, which is atomic but requires the path to be on the same filesystem.
+
+Atomicity is important to ensure that `$cas` entries are always valid. If they are copied instead, they don't self-validate for the duration of the copy.
+
+#### with Store Daemon
+
+Any user with write access to `/var/lib/nix/.prepare` and `/var/lib/nix/.stage` can ask for entries to be installed. To do so:
+
+1. They prepare entries to be stored in `/var/lib/nix/.prepare`, renaming each entry to its `$cas` and `$cas.m`.
+1. They atomically move prepared paths to `/var/lib/nix/.stage`, in reverse dependency order, meaning dependencies of an entry are moved first. First the `$cas.m` file is moved and then the `$cas` entry.
+
+When the Store daemon discovers a new `$cas` entry under `.stage`:
```

Comment:
Why not just do this process in `.prepare`? It can even be a setuid executable that you call to start the process. The nix-daemon then doesn't need to be constantly running and watching `.stage` and we can remove the need for `.daemon`. You just call the process when you are done with the path and it takes care of normalizing the directory it and then moves it into the store when done.

Of course the setuid executable isn't required if you have permission like the "without Store Daemon" approach. In that case you could still call the same binary (and reuse the security sensitive code) without the setuid bit.

To prevent DoS the "daemon" can require that the owner chmods the directory to the daemon user first or writes some magic cookie to the root to indicate that the directory is ready.